### PR TITLE
handle inlined function in hybrid

### PR DESCRIPTION
**Pull request by romainfrancois**
So that: 

```r
> summarise(mtcars, cyl = (!!sum)(cyl))
  cyl
1 198
> hybrid_call(mtcars, (!!sum)(cyl))
<hybrid evaluation>
  call      : .Primitive("sum")(cyl)
  C++ class : dplyr::hybrid::internal::SumTemplate<14, false, dplyr::NaturalDataFrame>
> summarise_all(mtcars, list(sum = sum))
  mpg_sum cyl_sum disp_sum hp_sum drat_sum  wt_sum qsec_sum vs_sum am_sum gear_sum carb_sum
1   642.9     198   7383.1   4694   115.09 102.952   571.16     14     13      118       90
```

Also, although not sure about it yet: 

```
> dplyr:::hybrids()
# A tibble: 21 x 3
   name       package fun      
   <chr>      <chr>   <list>   
 1 sd         stats   <fn>     
 2 var        stats   <fn>     
 3 %in%       base    <fn>     
 4 max        base    <builtin>
 5 min        base    <builtin>
 6 mean       base    <fn>     
 7 sum        base    <builtin>
 8 n_distinct dplyr   <fn>     
 9 lag        dplyr   <fn>     
10 cume_dist  dplyr   <fn>     
# ... with 11 more rows
```

### CODE REVIEW

**Comment by romainfrancois** [comment]
This is related to the problems we have about hybridizing and `funs`. 

With this, we would no longer have to use the name of the function, we can use the function itself and hybrid internals match that against a hash map.

File: inst/include/dplyr/hybrid/Expression.h [hunk 1]
```diff
@@ -87,6 +98,13 @@ class Expression {
       // a symbol
       valid = true;
       func = head;
+    } else if (TYPEOF(head) == CLOSXP || TYPEOF(head) == BUILTINSXP) {
```
Comment (hadley) [inline inst/include/dplyr/hybrid/Expression.h]: of `TYPEOF(head) == SPECIALSXP` ?

Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: Oh yeah i guess %in% would be a special

**Comment by romainfrancois** [comment]
I think the mean <- sum already worked because we look what the symbol is and fall back to r if it’s not what we expect. 

What we gain though is that now it would use hybrid evaluation

**Comment by hadley** [comment]
Yeah, I think so. That way these both would use hybrid:

```R
mean <- sum
sum1 <- sum

dplyr::summarise(data.frame(x = 1:10), mean(x))
dplyr::summarise(data.frame(x = 1:10), sum1(x))
```

That does mean we need to evaluate the first part of the function call though.

We'd need to think about weirder situations like:

```R
f <- function() mean
x <- list(f = mean)

dplyr::summarise(data.frame(x = 1:10), f()(x))
dplyr::summarise(data.frame(x = 1:10), x$f(x))
```

**Comment by romainfrancois** [comment]
No longer based on the name but on what it resolves to. We handle 3 cases: 

- `pkg::fun(...)` : trivial
- `<symbol>(...)` : resolve the symbol and match against the hash map
- `<function>(...)` : match against the hash map

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
mean <- sum
hybrid_call(data.frame(x = 1:10), mean(x))
#> <hybrid evaluation>
#>   call      : base::sum(x)
#>   C++ class : dplyr::hybrid::internal::SumTemplate<13, false, dplyr::NaturalDataFrame>

mean <- function(...) 42
hybrid_call(data.frame(x = 1:10), mean(x))
#> <standard evaluation>
#>   call      : mean(x)
hybrid_call(data.frame(x = 1:10), base::mean(x))
#> <hybrid evaluation>
#>   call      : base::mean(x)
#>   C++ class : dplyr::hybrid::internal::SimpleDispatchImpl<13, false, dplyr::NaturalDataFrame, dplyr::hybrid::internal::MeanImpl>
hybrid_call(data.frame(x = 1:10), (!!base::mean)(x))
#> <hybrid evaluation>
#>   call      : base::mean(x)
#>   C++ class : dplyr::hybrid::internal::SimpleDispatchImpl<13, false, dplyr::NaturalDataFrame, dplyr::hybrid::internal::MeanImpl>

rm(mean)
hybrid_call(data.frame(x = 1:10), mean(x))
#> <hybrid evaluation>
#>   call      : base::mean(x)
#>   C++ class : dplyr::hybrid::internal::SimpleDispatchImpl<13, false, dplyr::NaturalDataFrame, dplyr::hybrid::internal::MeanImpl>
```

Created on 2018-10-04 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)

**Comment by romainfrancois** [comment]
This case is handled with standard evaluation: 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

f <- function() mean
summarise(data.frame(x=1:10), f()(x))
#>   f()(x)
#> 1    5.5
hybrid_call(data.frame(x=1:10), f()(x))
#> <standard evaluation>
#>   call      : f()(x)
```

i.e. we don't eval `f()`

**Comment by romainfrancois** [comment]
This also goes through standard evaluation, and therefore fails because `x` from the data has priority. That's fine I think.  

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

x <- list(f = mean)
hybrid_call(data.frame(x=1:10), x$f(x))
#> <standard evaluation>
#>   call      : x$f(x)
summarise(data.frame(x=1:10), x$f(x))
#> Error: Evaluation error: $ operator is invalid for atomic vectors.
```

Created on 2018-10-04 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)

**Comment by hadley** [comment]
That all seems reasonable to me. I think you should get a review of the C++ from @lionel- or @krlmlr; I'm happy with the basic idea and high-level implementation.

File: tests/testthat/helper-hybrid.R [hunk 2]
```diff
@@ -1,9 +1,7 @@
 expect_hybrid <- function(data, expr, info = NULL, label = NULL) {
-  expr <- enquo(expr)
-  expect_true(hybrid_call(data, !!expr), info = info, label = label)
+  expect_true(hybrid_impl(data, enquo(expr)), info = info, label = label)
```
Comment (krlmlr) [inline tests/testthat/helper-hybrid.R]: Do we need `!!enquo(expr)` here?

File: src/hybrid.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      hybrid_function(name, package, id)
+    )
+  );
+}
+
+}
+}
+
+// [[Rcpp::init]]
+void init_hybrid_inline_map(DllInfo* dll) {
```
Comment (krlmlr) [inline src/hybrid.cpp]: If this method doesn't use `dll`, compiler warnings can be suppressed with a `(void)dll;` or an `UNUSED` macro.

File: inst/include/dplyr/hybrid/scalar_result/mean_sd_var.h [hunk 4]
```diff
@@ -183,6 +184,82 @@ SEXP sd_(const SlicedTibble& data, Column variable, bool narm, const Operation&
...
+  Column x;
+
+  switch (expression.size()) {
+  case 1:
+    // mean( <column> )
+    if (expression.is_unnamed(0) && expression.is_column(0, x)) {
+      return mean_(data, x, false, op);
+    }
+  case 2:
+    bool test;
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/scalar_result/mean_sd_var.h]: Here and further below: Should this be named `na_rm` and perhaps initialized as `false` outside the `switch()` and used also for `case 1` ?

File: inst/include/dplyr/hybrid/hybrid.h [hunk 5]
```diff
@@ -28,319 +28,62 @@ template <typename SlicedTibble, typename Operation>
...
-  Environment ns_stats = Environment::namespace_env("stats");
-
   Expression<SlicedTibble> expression(expr, mask, env);
-  if (!expression.is_valid()) return R_UnboundValue;
-
-  // functions that take variadic number of arguments
-  if (expression.is_fun(symbols::n_distinct, symbols::dplyr, ns_dplyr)) {
+  switch (expression.get_id()) {
+  case N_DISTINCT:
     return n_distinct_(data, expression, op);
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/hybrid.h]: Should this be `n_distinct_dispatch()` for consistency?

File: inst/include/dplyr/hybrid/hybrid.h [hunk 6]
```diff
@@ -28,319 +28,62 @@ template <typename SlicedTibble, typename Operation>
...
-        return nth3_default(data, x, n, expression.value(2), op);
-      }
-    }
-
-    // give up
-    return R_UnboundValue;
-
-  default:
+  case N:
+    return expression.size() == 0 ? op(n_(data)) : R_UnboundValue;
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/hybrid.h]: Extracting trivial `n_dispatch()` and `group_indices_dispatch()` functions for a completely homogeneous `switch()` would help readability.

File: tests/testthat/helper-hybrid.R (hunk 2 above)
Comment (romainfrancois) [reply tests/testthat/helper-hybrid.R]: 🤔 it needs a quosure

File: inst/include/dplyr/hybrid/hybrid.h (hunk 5 above)
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/hybrid.h]: Yes. I'll make it more consistent.

File: inst/include/dplyr/hybrid/hybrid.h (hunk 6 above)
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/hybrid.h]: yes

File: tests/testthat/helper-hybrid.R (hunk 2 above)
Comment (krlmlr) [reply tests/testthat/helper-hybrid.R]: I don't understand. I can't find `hybrid_impl` nor `hybrid_call` in the diff, how did the code work before? See also https://github.com/tidyverse/dplyr/pull/3876/files#diff-26dc2875c1d102cbdcd1041fc6ebb1a8R747.

Also, `expect_not_hybrid()` still uses `hybrid_call()` .

**Comment by lionel-** [comment]
it would also be nice to handle simple functions:

```r
fn <- function(x) mean(x, na.rm = TRUE)
summarise(df, fn(var))
```

This would help with hybrid evaluation and colwise functions specified with the formula shortcut:

```r
summarise_at(df, "var", ~mean(.x))
```

**Comment by romainfrancois** [comment]
🤔 I'm not sure how to handle this: 

```r
fn <- function(x) mean(x, na.rm = TRUE)
summarise(df, fn(var))
```

but can the `~` mark the function somehow ?

**Comment by romainfrancois** [comment]
Related to #3843

**Comment by hadley** [comment]
That last case seems like it's out of scope for this PR since it will need considerably more spelunking into the function body.

File: inst/include/dplyr/hybrid/scalar_result/mean_sd_var.h (hunk 4 above)
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/scalar_result/mean_sd_var.h]: ✅

**Pull request merged**